pkg/relay/register.go

```diff
@@ -0,0 +1,296 @@
+package relay
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/blocknative/dreamboat/pkg/structs"
+	"github.com/flashbots/go-boost-utils/types"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	ResponseTypeVerify = iota
+	ResponseTypeOthers
+	ResponseTypeStored
+)
+
+type TimestampRegistry interface {
+	Get(pubkey string) (timestamp uint64, ok bool)
+}
+
+// VerifyReq is a request structure used in communication
+// between api calls and fixed set of worker goroutines
+// it's using return channel pattern, meaning that after
+// sent the sender locks on that channel to get the response
+type VerifyReq struct {
+	Signature [96]byte
+	Pubkey    [48]byte
+	Msg       [32]byte
+	// Unique identifier of payload
+	// if needed to be passed back in response
+	ID       int
+	Response chan Resp
+}
+
+// StoreReq is similar to VerifyReq jsut for storing payloads
+type StoreReq struct {
+	RawPayload json.RawMessage
+	Pubkey     types.PublicKey
+	ID         int
+	Response   chan Resp
+}
+
+// Resp respone structure
+// - potential candidate for structure pool
+// as it's almost constant size
+type Resp struct {
+	ID     int
+	Type   int8
+	Commit bool
+	Err    error
+}
+
+// ***** Builder Domain *****
+// RegisterValidator is called is called by validators communicating through mev-boost who would like to receive a block from us when their slot is scheduled
+func (rs *Relay) RegisterValidator(ctx context.Context, payload []structs.SignedValidatorRegistration) error {
+	timer := prometheus.NewTimer(rs.m.Timing.WithLabelValues("registerValidator", "all"))
+	defer timer.ObserveDuration()
+
+	var respCh chan Resp
+
+	// This function is limitted to the size of the buffer to prevent deadlocks
+	if rs.config.RegisterValidatorMaxNum > uint64(len(payload)) {
+		respCh = rs.retChannPool.Get().(chan Resp)
+		defer rs.retChannPool.Put(respCh)
+	}
+
+	fc := NewFlowControl(respCh, len(payload))
+	defer fc.Close()
+
+	// verify other parameters in separate goroutine
+	go verifyOthers(rs.beaconState, rs.regMngr, fc, payload)
+
+	// synchronize responses from:
+	//	- verification of signatures
+	//	- other verifications of payload validity
+	//	- successfull store
+	go synchAndStore(rs.regMngr, fc, payload)
+
+	verifyChan := rs.regMngr.GetVerifyChan(ResponseQueueRegister)
+SendPayloads:
+	for i, p := range payload {
+		msg, err := types.ComputeSigningRoot(payload[i].Message, rs.config.BuilderSigningDomain)
+		if err != nil {
+			fc.SentVerificationInc()
+			fc.RespCh <- Resp{ID: i, Err: errors.New("invalid signature"), Type: ResponseTypeVerify}
+			break SendPayloads
+		}
+
+		select {
```

Comment:
Can we put this into a function and return a bool that determines whether to break out of the loop?  Labeled breaks are hard to follow, and the statefulness tends to cause bugs.

Reply:
I'm not exactly following what kind of change are you talking bout here :) 
This operation has to be a channel select.

Reply:
I'm proposing that we wrap the entire select statement in its own function, which would return a bool that indicates whether to break out of the loop.  The idea is to get rid of the `SendPayloads` label.

Reply:
I don't like labels 😉